fabtest/rdm_stress: Add stress test for RDM endpoints

shefty opened the pull request:
Stress multi-threaded and multi-process access to an RDM endpoint
with abrupt termination conditions.  This will help verify or
expose race conditions in error handling cases.

This patch defines the general framework of operation and
implements most of the client code.  The server side is still
needed.

The general operation is for the client to parse an input file
describing a set of RPCs to perform.  Each RPC involves sending
a short message to the server.  The server will respond by either
sending a message, a tagged message, or issuing an RMA read or
write operation back to the client.  RMA operations are followed
by a response message.

The client side has a parent process and several child processes.
Child processes are forked from the parent, with each child
executing a set of RPC calls to a single server.  The parent
spawns new children as children processes exit.  A child process
may not wait for an RPC response and could exit after sending
a request.

The intent for the server is that the main thread will receive
RPC messages and spawn a thread to handle that request.  The
new thread will generate the RPC response and exit.

The objective is to test a multi-threaded server under load
from multiple clients which are possibly misbehaving.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

j-xiong commented on fabtests/functional/rdm_stress.c:
```diff
@@ -0,0 +1,479 @@
+	for (i = 0; i < opts.iterations; i++) {
+		for (myid = 0; myid < opts.num_connections; myid++) {
+			if (clients[myid].pid) {
+				waitpid(clients[myid].pid, NULL, 0);
```
Would it be better to use wait() or waitpid(-1, ..) to catch any child process instead wait for them in turn?

shefty commented on fabtests/functional/rdm_stress.c:
```diff
@@ -0,0 +1,479 @@
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_tagged.h>
+#include <shared.h>
```
For fabtests, these are installed headers.

shefty replied:
Oh - I think you're only referring to shared.h.  I'll update

j-xiong replied:
Yes, only the single line.

shefty commented:
Added 2 patches to the av implementation.  The first adds a print if a duplicate address is inserted.  The second restructures the rxm av remove code to optimize the locking and avoid issues if a duplicate address is inserted (depending on the app's behavior).  Those patches are both candidates to pull into v1.15.

The rdm_stress program is now working, with the exception that the input file parsing is commented out.  Instead I have a hard-coded 'input' array that I am using to test and debug the framework.  That array will be removed once the parsing code is re-enabled.  I also changed the input format (simplified it), so the parsing code needs to be updated.

The good news is that the stress program appears to be doing the job of exposing issues.  The bad news is that I don't know if the issues are in the program or libfabric, but I believe the latter.
